In [1]:
import heapq

# 计算每个烧杯的期望值函数
def calculate_expectation(l, b, p):
    return (b / (l + b)) * p

# 初始化
n, t, q = map(int, input().split())  # n个烧杯，t毫升物质β，q次变动
p = list(map(int, input().split()))  # 每个烧杯的pi值
l = list(map(int, input().split()))  # 每个烧杯的li值

# 当前最大期望值的优先队列（贪心策略）
pq = []

# 初始状态下的期望值
total_expectation = 0
for i in range(n):
    current_expectation = calculate_expectation(l[i], min(l[i], t), p[i])
    heapq.heappush(pq, (-current_expectation, i))  # 用负数来实现最大堆
    total_expectation += current_expectation

# 处理每次变动
for _ in range(q):
    tj, rj = map(int, input().split())  # 读取操作
    rj -= 1  # 将烧杯编号转换为0-indexed

    # 对应操作
    if tj == 1:
        l[rj] += 1  # 增加物质α
    elif tj == 2:
        l[rj] -= 1  # 减少物质α
    
    # 更新该烧杯的期望值
    current_expectation = calculate_expectation(l[rj], min(l[rj], t), p[rj])
    
    # 总期望值更新
    heapq.heappush(pq, (-current_expectation, rj))
    
    # 输出当前期望值
    print(f"{current_expectation:.9f}")

1.333333333
1.250000000
2.000000000


In [2]:
import heapq

# 计算每个烧杯的期望值函数
def calculate_expectation(l, b, p):
    return (b / (l + b)) * p

n, t, q = map(int, input().split())  # n个烧杯，t毫升物质β，q次变动
p = list(map(int, input().split()))  # 每个烧杯的pi值
l = list(map(int, input().split()))  # 每个烧杯的li值

# 当前最大期望值的优先队列（贪心策略）
pq = []
for i in range(n):
    pq.append((-calculate_expectation(l[i], min(l[i], t), p[i]), i))

# 转成堆
heapq.heapify(pq)

# 计算初始的总期望值
total_expectation = 0
for i in range(n):
    total_expectation += calculate_expectation(l[i], min(l[i], t), p[i])

# 处理每次变动
for _ in range(q):
    tj, rj = map(int, input().split())  # 读取操作
    rj -= 1  # 将烧杯编号转换为0-indexed

    # 对应操作
    if tj == 1:
        l[rj] += 1  # 增加物质α
    elif tj == 2:
        l[rj] -= 1  # 减少物质α
    
    # 更新该烧杯的期望值
    heapq.heappush(pq, (-calculate_expectation(l[rj], min(l[rj], t), p[rj]), rj))

    # 计算总期望值
    total_expectation = 0
    for i in range(n):
        total_expectation += calculate_expectation(l[i], min(l[i], t), p[i])
    
    # 输出当前总期望值
    print(f"{total_expectation:.9f}")

3.000000000
2.583333333
3.250000000


In [1]:
import heapq

def main():
    n, t, q = map(int, input().split())
    p = list(map(int, input().split()))
    l = list(map(int, input().split()))
    
    # Initialize b_i = 0 for all containers
    b = [0] * n
    
    # Compute initial marginal gains and build a max heap
    # Since heapq is a min heap, we use negative gains
    heap = []
    for i in range(n):
        if l[i] == 0:
            gain = 0
        else:
            gain = p[i] / (l[i] + 1)
        heap.append((-gain, i))
    
    heapq.heapify(heap)
    
    # Allocate t units
    for _ in range(t):
        if not heap:
            break
        neg_gain, i = heapq.heappop(heap)
        # Assign one unit to container i
        b[i] += 1
        if b[i] < l[i]:
            # Compute new marginal gain
            gain = p[i] * l[i] / (l[i] + b[i] + 1)**2
            heapq.heappush(heap, (-gain, i))
    
    # Compute initial expected value
    total_E = 0.0
    for i in range(n):
        if b[i] > 0:
            total_E += (b[i] / (l[i] + b[i])) * p[i]
    
    # Process queries
    for _ in range(q):
        tj, rj = map(int, input().split())
        rj -= 1  # 0-based index
        # Update l[rj]
        if tj == 1:
            l[rj] += 1
        elif tj == 2:
            l[rj] -= 1
        
        # Recompute the allocation
        # Reset allocations
        b = [0] * n
        heap = []
        for i in range(n):
            if l[i] == 0:
                gain = 0
            else:
                gain = p[i] / (l[i] + 1)
            heap.append((-gain, i))
        heapq.heapify(heap)
        # Allocate t units
        for _ in range(t):
            if not heap:
                break
            neg_gain, i = heapq.heappop(heap)
            # Assign one unit to container i
            b[i] += 1
            if b[i] < l[i]:
                # Compute new marginal gain
                gain = p[i] * l[i] / (l[i] + b[i] + 1)**2
                heapq.heappush(heap, (-gain, i))
        # Compute expected value
        total_E = 0.0
        for i in range(n):
            if b[i] > 0:
                total_E += (b[i] / (l[i] + b[i])) * p[i]
        # Print the result with 9 decimal places
        print("{0:.9f}".format(total_E))

if __name__ == "__main__":
    main()

1.666666667
1.333333333
2.000000000


### 题目理解

科研人员有 `n` 个烧杯，每个烧杯中含有一定量的物质 α (`l_i` 毫升)。他们拥有 `t` 毫升的物质 β，可以分配到这些烧杯中。每个烧杯中加入 `b_i` 毫升物质 β 后，研制成功的概率为：

\[
\text{概率} = \frac{b_i}{l_i + b_i}
\]

研制成功后，每个烧杯可以获得 `p_i` 毫升的新能源。目标是合理分配物质 β，使得新能源的期望值最大。需要注意的是，每个烧杯中物质 β 的量不能超过物质 α 的量 (`b_i ≤ l_i`)，否则会发生爆炸。

此外，烧杯中物质 α 的量会发生 `q` 次变动（增加或减少 1 毫升），每次变动后需要重新计算最大期望值。

### 样例分析

**输入说明：**
```
2 1 3 
4 5 
1 2 
1 1 
1 2 
2 1 
```

- `n = 2` 个烧杯
- `t = 1` 毫升物质 β
- `q = 3` 次变动

**初始状态：**
- `p = [4, 5]`
- `l = [1, 2]`

**变动过程及对应输出：**

1. **第一次变动：`1 1`**
   - **操作**：`tj = 1` 表示增加 1 毫升物质 α，`rj = 1` 表示第一个烧杯。
   - **更新后**：`l = [2, 2]`
   
   **分配方案分析：**
   
   可分配的物质 β 总量为 1 毫升，且每个烧杯中 `b_i ≤ l_i`。
   
   - **方案1**：将 1 毫升 β 分配给烧杯1 (`b1 = 1`, `b2 = 0`)
     \[
     \text{期望值} = \frac{1}{2 + 1} \times 4 + \frac{0}{2 + 0} \times 5 = \frac{1}{3} \times 4 + 0 = \frac{4}{3} \approx 1.333333333
     \]
   
   - **方案2**：将 1 毫升 β 分配给烧杯2 (`b1 = 0`, `b2 = 1`)
     \[
     \text{期望值} = \frac{0}{2 + 0} \times 4 + \frac{1}{2 + 1} \times 5 = 0 + \frac{1}{3} \times 5 = \frac{5}{3} \approx 1.666666667
     \]
   
   **最佳方案**：方案2，期望值为 `1.666666667`

2. **第二次变动：`1 2`**
   - **操作**：`tj = 1` 表示增加 1 毫升物质 α，`rj = 2` 表示第二个烧杯。
   - **更新后**：`l = [2, 3]`
   
   **分配方案分析：**
   
   - **方案1**：将 1 毫升 β 分配给烧杯1 (`b1 = 1`, `b2 = 0`)
     \[
     \text{期望值} = \frac{1}{2 + 1} \times 4 + \frac{0}{3 + 0} \times 5 = \frac{1}{3} \times 4 + 0 = \frac{4}{3} \approx 1.333333333
     \]
   
   - **方案2**：将 1 毫升 β 分配给烧杯2 (`b1 = 0`, `b2 = 1`)
     \[
     \text{期望值} = \frac{0}{2 + 0} \times 4 + \frac{1}{3 + 1} \times 5 = 0 + \frac{1}{4} \times 5 = \frac{5}{4} = 1.25
     \]
   
   **最佳方案**：方案1，期望值为 `1.333333333`

3. **第三次变动：`2 1`**
   - **操作**：`tj = 2` 表示减少 1 毫升物质 α，`rj = 1` 表示第一个烧杯。
   - **更新后**：`l = [1, 3]`
   
   **分配方案分析：**
   
   - **方案1**：将 1 毫升 β 分配给烧杯1 (`b1 = 1`, `b2 = 0`)
     \[
     \text{期望值} = \frac{1}{1 + 1} \times 4 + \frac{0}{3 + 0} \times 5 = \frac{1}{2} \times 4 + 0 = 2.0
     \]
   
   - **方案2**：将 1 毫升 β 分配给烧杯2 (`b1 = 0`, `b2 = 1`)
     \[
     \text{期望值} = \frac{0}{1 + 0} \times 4 + \frac{1}{3 + 1} \times 5 = 0 + \frac{1}{4} \times 5 = 1.25
     \]
   
   **最佳方案**：方案1，期望值为 `2.000000000`

**最终输出：**
```
1.666666667 
1.333333333 
2.000000000 
```

### 总结

每次变动后，通过比较将物质 β 分配到不同烧杯所能获得的期望值，选择期望值最大的分配方案。上述样例中，经过三次变动后，分别得到期望值 `1.666666667`、`1.333333333` 和 `2.000000000`。

### 解题思路

这道题目要求在多次动态变化下，合理分配有限的物质 β 以最大化新能源的期望值。由于输入规模较大（`n`, `t`, `q` 均可达 200,000），需要设计高效的算法来处理。以下是解决该问题的思路：

#### 1. **理解期望值的最大化目标**

每个烧杯分配 `b_i` 毫升的物质 β 后，研制成功的概率为：

$$
\text{概率} = \frac{b_i}{l_i + b_i}
$$

成功后得到的新能源量为 `p_i` 毫升，因此该烧杯的期望新能源量为：

$$
E_i = \frac{b_i}{l_i + b_i} \times p_i
$$

目标是分配 `t` 毫升的物质 β，使得所有烧杯的期望新能源量之和最大：

$$
\text{最大化} \quad \sum_{i=1}^{n} \frac{b_i}{l_i + b_i} \times p_i
$$

同时，需满足：

$$
0 \leq b_i \leq l_i \quad \text{且} \quad \sum_{i=1}^{n} b_i = t
$$

#### 2. **分配策略的贪心原则**

由于每分配一个单位的 β 到某个烧杯，其带来的边际期望增益是不同的，且随着分配量的增加，边际增益是递减的, 因为：

$$
\frac{dE_i}{db_i} = \frac{p_i l_i}{(l_i + b_i)^2}
$$

因此，最佳的分配策略应遵循贪心原则：每次将一个单位的 β 分配给当前边际增益最高的烧杯。

#### 3. **边际增益的计算与优先级队列的应用**

为了高效地选择当前边际增益最高的烧杯，可以采用优先级队列（最大堆）的数据结构。具体步骤如下：

1. **初始准备**：
   - 对于每个烧杯，初始化 `b_i = 0`。
   - 计算每个烧杯当前的边际增益：

     $$
     \text{Marginal Gain} = \frac{p_i \times l_i}{(l_i + b_i)^2} = \frac{p_i \times l_i}{(l_i)^2} = \frac{p_i}{l_i}
     $$

   - 将所有烧杯的初始边际增益及其相关信息（如 `i`、当前 `b_i`）加入最大堆。

2. **分配物质 β**：
   - 重复 `t` 次以下步骤：
     - 弹出堆顶元素，即当前边际增益最高的烧杯。
     - 将该烧杯的 `b_i` 增加 1（注意不能超过 `l_i`）。
     - 重新计算该烧杯的新边际增益，并将其重新加入堆中。

3. **处理动态变化**：
   - 每次烧杯中物质 α (`l_i`) 发生变化后，需要更新相关烧杯的边际增益。
   - 由于 `l_i` 变化影响到边际增益的计算，因此需要重新调整该烧杯在堆中的位置。
   - 具体步骤：
     - 更新对应烧杯的 `l_i`（增加或减少 1 毫升）。
     - 重新计算其当前的边际增益。
     - 更新堆中该烧杯的位置，确保堆的性质保持（即堆顶始终是当前最高的边际增益）。

4. **优化考虑**：
   - 由于直接在每次变动后重新分配 `t` 个 β 可能导致时间复杂度过高（尤其是 `t` 和 `q` 都很大时），需要进行优化。
   - **动态规划** 或 **二分查找** 等方法可能需要结合具体问题的特性进行设计，但在本题中，贪心加堆的方式更为直接且高效。
   - 进一步优化可以通过**分批处理**或**延迟更新**等技术减少不必要的计算。

#### 4. **数据结构与时间复杂度分析**

- **优先级队列（最大堆）**：用于高效地获取当前边际增益最高的烧杯。
- **时间复杂度**：
  - 初始化堆的时间复杂度为 `O(n)`。
  - 每次分配 β 的操作涉及 `O(log n)` 的堆操作，若 `t` 很大，需考虑批量分配或其他优化策略。
  - 每次变动后的更新也涉及 `O(log n)` 的堆操作。
  - 总体时间复杂度在可接受范围内，但需要精细优化以应对最大输入规模。

#### 5. **处理边界条件**

- **爆炸条件**：确保分配的 `b_i` 不超过当前 `l_i`。在分配前需要检查是否已达到上限。
- **动态变化的有效性**：保证每次变动后，`l_i` 不会导致任何烧杯中物质 α 量为零或负数（题目已保证所有烧杯不为空）。
- **精度要求**：由于输出需要高精度（相对或绝对误差不超过 `1e-6`），在计算期望值时需要使用高精度的数据类型（如双精度浮点数）。

### 总结

通过将问题转化为在每次动态变化后，贪心地分配物质 β 到当前边际增益最高的烧杯，可以有效地最大化新能源的期望值。利用优先级队列来管理和更新边际增益，结合高效的堆操作，可以在大规模输入下实现快速计算。关键在于理解边际增益的递减特性，并设计合适的数据结构来支持动态更新。